iothub/service/src/Messaging/FileUploadNotificationProcessorClient.cs

```diff
@@ -0,0 +1,215 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Net.Sockets;
+using System.Net.WebSockets;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.Amqp;
+using Microsoft.Azure.Devices.Common;
+using Microsoft.Azure.Devices.Common.Exceptions;
+using Microsoft.Azure.Devices.Common.Extensions;
+
+namespace Microsoft.Azure.Devices
+{
+    /// <summary>
+    /// Subclient of <see cref="IotHubServiceClient"/> for handling file upload notications.
+    /// </summary>
+    /// <seealso href="https://docs.microsoft.com/azure/iot-hub/iot-hub-devguide-file-upload#service-file-upload-notifications"/>.
+    public class FileUploadNotificationProcessorClient : IDisposable
+    {
+        private readonly string _hostName;
+        private readonly IotHubConnectionProperties _credentialProvider;
+        private readonly IotHubConnection _connection;
+
+        /// <summary>
+        /// The callback to be executed each time file upload notification is received from the service.
+        /// </summary>
+        /// <remarks>
+        /// May not be null.
+        /// </remarks>
+        public Func<FileNotification, DeliveryAcknowledgement> _fileNotificationProcessor;
```

Comment:
This ```DeliveryAcknowledgement``` type isn't what we need from the user here. We need an enum with three possible values:

```csharp
public enum AcknowledgementType
{
    Complete, // maps to Completed in AMQP
    Abandon, // maps to Released in AMQP
    Reject // maps to Rejected in AMQP
}
```

Reply:
Complete means that the received file upload notification/feedback message was received and handled and the service can remove it from the queue. 

Abandon means the received file upload notifications/feedback message was received but not handled and that the service should requeue it so that this or any other receiver can receive it later.

Rejected means the file upload notification was received but was not handled because it was poorly formatted or otherwise unhandleable. This acknowledgement type makes the service remove the item from the queue like with Completed.

Reply:
It is very similar to how device clients can abandon/reject/complete cloud to device messages, so [this reading](https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-messages-c2d#the-cloud-to-device-message-life-cycle) may be helpful

Reply:
Although, now that I look at the Java implementation, the service may not support Reject here.

https://github.com/Azure/azure-iot-sdk-java/blob/main/service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/messaging/AcknowledgementType.java